src/Microsoft.TemplateEngine.Edge/GlobalSettingsTemplatePackages/GlobalSettingsTemplatePackagesProvider.cs

```diff
@@ -0,0 +1,250 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Abstractions.GlobalSettings;
+using Microsoft.TemplateEngine.Abstractions.Installer;
+using Microsoft.TemplateEngine.Abstractions.TemplatePackages;
+using Microsoft.TemplateEngine.Utils;
+
+namespace Microsoft.TemplateEngine.Edge
+{
+    internal class GlobalSettingsTemplatePackagesProvider : IManagedTemplatePackagesProvider
+    {
+        private readonly string PackagesFolder;
+        private IEngineEnvironmentSettings _environmentSettings;
+        private Dictionary<Guid, IInstaller> _installersByGuid = new Dictionary<Guid, IInstaller>();
+        private Dictionary<string, IInstaller> _installersByName = new Dictionary<string, IInstaller>();
+
+        public GlobalSettingsTemplatePackagesProvider
+            (GlobalSettingsTemplatePackagesProviderFactory factory, IEngineEnvironmentSettings settings)
+        {
+            _ = factory ?? throw new ArgumentNullException(nameof(factory));
+            _ = settings ?? throw new ArgumentNullException(nameof(settings));
+
+            Factory = factory;
+            PackagesFolder = Path.Combine(settings.Paths.TemplateEngineRootDir, "packages");
+            if (!settings.Host.FileSystem.DirectoryExists(PackagesFolder))
+            {
+                settings.Host.FileSystem.CreateDirectory(PackagesFolder);
+            }
+
+            _environmentSettings = settings;
+            foreach (var installerFactory in settings.SettingsLoader.Components.OfType<IInstallerFactory>())
+            {
+                var installer = installerFactory.CreateInstaller(this, settings, PackagesFolder);
+                _installersByName[installerFactory.Name] = installer;
+                _installersByGuid[installerFactory.Id] = installer;
+            }
+
+            // We can't just add "SettingsChanged+=SourcesChanged", because SourcesChanged is null at this time.
+            settings.SettingsLoader.GlobalSettings.SettingsChanged += () => SourcesChanged?.Invoke();
+        }
+
+        public event Action SourcesChanged;
+
+        public ITemplatePackagesProviderFactory Factory { get; }
+
+        public async Task<IReadOnlyList<ITemplatePackage>> GetAllSourcesAsync(CancellationToken cancellationToken)
+        {
+            var list = new List<ITemplatePackage>();
+            foreach (TemplatePackageData entry in await _environmentSettings.SettingsLoader.GlobalSettings.GetInstalledTemplatesPackagesAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (_installersByGuid.TryGetValue(entry.InstallerId, out var installer))
+                {
+                    list.Add(installer.Deserialize(this, entry));
```

Comment:
I think installers should throw when they cannot deserialize something, i.e. mandatory information is not there, for example identifier for NuGet etc. 
Settings fie can be modified externally, and we better skip invalid entries at the beginning then try to process and then fail afterwards.